### DISCUSSION TITLE
custom file for eMylo 3-Phase Smart Wifi Energy Meter

### DISCUSSION THREAD

**Discussion by dnwalker68** [make-all/tuya-local#2320] [Q&A]
Hello everyone,
unfortunately I have absolutely no idea in this area. I read the documentation on how to write such a file, but I didn't really understand it. So don't kill me if I ask a stupid question.

I have a Make eMylo smart meter here. However, it is not the 1-phase device that is already in the device database but a 3-phase device. Here is a link to the device on Amazon.

[https://www.amazon.de/dp/B0D4DT8V37?tag=operapc-def-sp-co2-de-21](url)

Here is an excerpt of the data from the Tuya Developer Tool

`ResponseCopy
{
  "result": [
    {
      "active_time": 1726762981,
      "bind_space_id": "61381298",
      "category": "dlq",
      "create_time": 1726762981,
      "custom_name": "",
      "icon": "smart/icon/ay1553162414592ZG6Kk/ed250f82f5259b55a100ffa430d9ef49.png",
      "id": "bf7ad3617f2502b02foktb",
      "ip": "xxx.xxx.xxx.xxx",
      "is_online": true,
      "lat": "51.01",
      "local_key": "xxxxxxxxxxxx",
      "lon": "6.89",
      "model": "",
      "name": "Metering_3PN_WiFi",
      "product_id": "fzxonzodwjiugioz",
      "product_name": "SDM01_WiFi+485_V1",
      "sub": false,
      "time_zone": "+02:00",
      "update_time": 1726762984,
      "uuid": "d35dcc51f1f47a5e"
    }
  ],
  "success": true,
  "t": 1726764507945,
  "tid": "fe89f5ca76a611ef88b02ac421ecbce6"
}

######################################################################

{
  "result": {
    "properties": [
      {
        "code": "forward_energy_total",
        "custom_name": "",
        "dp_id": 1,
        "time": 1726763247371,
        "type": "value",
        "value": 30
      },
      {
        "code": "phase_a",
        "custom_name": "",
        "dp_id": 6,
        "time": 1726764314860,
        "type": "raw",
        "value": "CQUAAAAAAAA="
      },
      {
        "code": "phase_b",
        "custom_name": "",
        "dp_id": 7,
        "time": 1726764315103,
        "type": "raw",
        "value": "CQkAAAAAAAA="
      },
      {
        "code": "phase_c",
        "custom_name": "",
        "dp_id": 8,
        "time": 1726764315232,
        "type": "raw",
        "value": "CQMAAAAAAAA="
      },
      {
        "code": "fault",
        "custom_name": "",
        "dp_id": 9,
        "time": 1726763247371,
        "type": "bitmap",
        "value": 0
      },
      {
        "code": "clear_energy",
        "custom_name": "",
        "dp_id": 12,
        "time": 1726763325450,
        "type": "bool",
        "value": true
      },
      {
        "code": "alarm_set_1",
        "custom_name": "",
        "dp_id": 17,
        "time": 1726763342221,
        "type": "raw",
        "value": "BwAAGQ=="
      },
      {
        "code": "alarm_set_2",
        "custom_name": "",
        "dp_id": 18,
        "time": 1726763342315,
        "type": "raw",
        "value": "AQAAZAIBAA8DAQD9BAAAtAUAAAAHAQAA"
      },
      {
        "code": "breaker_id",
        "custom_name": "",
        "dp_id": 19,
        "time": 1726763325714,
        "type": "string",
        "value": "fzx24010001"
      },
      {
        "code": "reverse_energy_total",
        "custom_name": "",
        "dp_id": 23,
        "time": 1726763325131,
        "type": "value",
        "value": 0
      },
      {
        "code": "active_power",
        "custom_name": "",
        "dp_id": 29,
        "time": 1726763743826,
        "type": "value",
        "value": 0
      },
      {
        "code": "supply_frequency",
        "custom_name": "",
        "dp_id": 32,
        "time": 1726764315274,
        "type": "value",
        "value": 5001
      },
      {
        "code": "online_state",
        "custom_name": "",
        "dp_id": 35,
        "time": 1726762980993,
        "type": "enum",
        "value": "online"
      },
      {
        "code": "power_factor",
        "custom_name": "",
        "dp_id": 50,
        "time": 1726763325435,
        "type": "value",
        "value": 99
      },
      {
        "code": "device_locating",
        "custom_name": "",
        "dp_id": 101,
        "time": 1726763325466,
        "type": "bool",
        "value": false
      },
      {
        "code": "data_report_duration",
        "custom_name": "",
        "dp_id": 102,
        "time": 1726763325479,
        "type": "value",
        "value": 15
      },
      {
        "code": "cur_voltage_x",
        "custom_name": "",
        "dp_id": 103,
        "time": 1726764315290,
        "type": "value",
        "value": 2309
      },
      {
        "code": "cur_current_x",
        "custom_name": "",
        "dp_id": 104,
        "time": 1726764315305,
        "type": "value",
        "value": 0
      },
      {
        "code": "cur_power_x",
        "custom_name": "",
        "dp_id": 105,
        "time": 1726764315326,
        "type": "value",
        "value": 0
      },
      {
        "code": "power_factor_x",
        "custom_name": "",
        "dp_id": 108,
        "time": 1726764315449,
        "type": "value",
        "value": 99
      },
      {
        "code": "x_total_import_energy",
        "custom_name": "",
        "dp_id": 109,
        "time": 1726764315492,
        "type": "value",
        "value": 10
      },
      {
        "code": "x_total_export_energy",
        "custom_name": "",
        "dp_id": 110,
        "time": 1726764318186,
        "type": "value",
        "value": 0
      },
      {
        "code": "cur_voltage_y",
        "custom_name": "",
        "dp_id": 112,
        "time": 1726764315328,
        "type": "value",
        "value": 2313
      },
      {
        "code": "cur_current_y",
        "custom_name": "",
        "dp_id": 113,
        "time": 1726764315338,
        "type": "value",
        "value": 0
      },
      {
        "code": "cur_power_y",
        "custom_name": "",
        "dp_id": 114,
        "time": 1726764315394,
        "type": "value",
        "value": 0
      },
      {
        "code": "power_factor_y",
        "custom_name": "",
        "dp_id": 117,
        "time": 1726764315465,
        "type": "value",
        "value": 99
      },
      {
        "code": "y_total_import_energy",
        "custom_name": "",
        "dp_id": 118,
        "time": 1726764315567,
        "type": "value",
        "value": 10
      },
      {
        "code": "y_total_export_energy",
        "custom_name": "",
        "dp_id": 119,
        "time": 1726764315583,
        "type": "value",
        "value": 0
      },
      {
        "code": "cur_voltage_z",
        "custom_name": "",
        "dp_id": 121,
        "time": 1726764315409,
        "type": "value",
        "value": 2307
      },
      {
        "code": "cur_current_z",
        "custom_name": "",
        "dp_id": 122,
        "time": 1726764315421,
        "type": "value",
        "value": 0
      },
      {
        "code": "cur_power_z",
        "custom_name": "",
        "dp_id": 123,
        "time": 1726764315436,
        "type": "value",
        "value": 0
      },
      {
        "code": "power_factor_z",
        "custom_name": "",
        "dp_id": 126,
        "time": 1726764315478,
        "type": "value",
        "value": 99
      },
      {
        "code": "z_total_import_energy",
        "custom_name": "",
        "dp_id": 127,
        "time": 1726764315593,
        "type": "value",
        "value": 10
      },
      {
        "code": "z_total_export_energy",
        "custom_name": "",
        "dp_id": 128,
        "time": 1726764315611,
        "type": "value",
        "value": 0
      },
      {
        "code": "current_unbalance",
        "custom_name": "",
        "dp_id": 130,
        "time": 1726764315622,
        "type": "value",
        "value": 0
      },
      {
        "code": "rs485_address",
        "custom_name": "",
        "dp_id": 131,
        "time": 1726763325520,
        "type": "value",
        "value": 247
      },
      {
        "code": "rs485_baut_rate",
        "custom_name": "",
        "dp_id": 132,
        "time": 1726763325536,
        "type": "enum",
        "value": "9600"
      },
      {
        "code": "rs485_data_format",
        "custom_name": "",
        "dp_id": 133,
        "time": 1726763325550,
        "type": "enum",
        "value": "N81"
      }
    ]
  },
  "success": true,
  "t": 1726764318959,
  "tid": "8de7f36a76a611efb77736033862a1ba"
}
`

I've now tried to create a config file.

[F_Tuya_Local_Metering_3PN_WiFi.yaml.pdf](https://github.com/user-attachments/files/17077889/F_Tuya_Local_Metering_3PN_WiFi.yaml.pdf)

Now when I try to install the device under HomeAssistant I get the following error in the log

Dieser Fehler wurde von einer benutzerdefinierten Integration verursacht

Logger: custom_components.tuya_local.helpers.device_config
Quelle: custom_components/tuya_local/helpers/device_config.py:998
Integration: Tuya Local (Dokumentation, Probleme)
Erstmals aufgetreten: 17:50:44 (1 Vorkommnisse)
Zuletzt protokolliert: 17:50:44

Parse error in Metering_3PN_WiFi.yaml

So there is an error somewhere in the file. Unfortunately I can't get any further now. It would be nice if you could help me here.
Please excuse my bad English.

Thank you and best regards

**Comment by dnwalker68** [comment]
Yes, it's a shame that no one can help...

**Comment by make-all** [comment]
I think very few people have the time to spend playing spot the syntax error in a PDF printout of someone else's YAML config. 

I suggest pasting your code into a validator such as https://yamlchecker.com/

**Reply by dnwalker68** [reply]
I have already run my code through various code checkers and no errors are displayed. I've already come up with the idea.

**Comment by dnwalker68** [comment]
Even when I try, I get the error message "Parse error in meter1.yaml"

`name: eMylo meter
products:
  - id: fzxonzodwjiugioz
    name: Metering-3PN-WiFi
primary_entity:
  entity: sensor
  class: energy
  dps:
    - id: 1
      type: value
      name: sensor
      unit: kWh
      class: total_increasing
    - id: 6
      type: raw
      name: phase_a
      optional: true
      force: true
    - id: 7
      type: raw
      name: phase_b
      optional: true
      force: true
    - id: 8
      type: raw
      name: phase_c
      optional: true
      force: true
secondary_entities:      
  - entity: sensor
    name: Current A
    class: current
    category: diagnostic
    dps:
      - id: 104
        type: value
        name: sensor
        unit: kW
        class: measurement
  - entity: sensor
    name: Current B
    class: current
    category: diagnostic
    dps:
      - id: 113
        type: value
        name: sensor
        unit: kW
        class: measurement
  - entity: sensor
    name: Current C
    class: current
    category: diagnostic
    dps:
      - id: 122
        type: value
        name: sensor
        unit: kW
        class: measurement
  - entity: sensor
    name: Voltage A
    class: voltage
    category: diagnostic
    dps:
      - id: 103
        type: value
        name: sensor
        unit: V
  - entity: sensor
    name: Voltage B
    class: voltage
    category: diagnostic
    dps:
      - id: 112
        type: value
        name: sensor
        unit: V
  - entity: sensor
    name: Voltage C
    class: voltage
    category: diagnostic
    dps:
      - id: 121
        type: value
        name: sensor
        unit: V
  - entity: sensor
    name: Power A
    class: power
    category: diagnostic
    dps:
      - id: 105
        type: value
        name: sensor
        unit: kW     
  - entity: sensor
    name: Power B
    class: power
    category: diagnostic
    dps:
      - id: 114
        type: value
        name: sensor
        unit: kW    
  - entity: sensor
    name: Power C
    class: power
    category: diagnostic
    dps:
      - id: 123
        type: value
        name: sensor
        unit: kW   
  - entity: sensor
    name: Total power
    class: power
    category: diagnostic
    dps:
      - id: 29
        type: value
        name: sensor
        unit: kW
        class: measurement
  - entity: button
    name: Clear_Energy
    category: config
    class: restart
    dps:
      - id: 12
        type: boolean
        name: button
        optional: true`

**Comment by dnwalker68** [comment]
Here I have the file again as a zip
[meter1.zip](https://github.com/user-attachments/files/17132209/meter1.zip)

**Reply by make-all** [reply]
"raw" and "value" are not valid types.

**Comment by sergt1978** [comment]
Please tell me, I'm new here, I'm mostly a product user, but not a developer. Perhaps you can tell me how and where to apply to expand the device's functions. As I wrote earlier, the message was hidden as unrelated to the topic. I understand that the topic is closed, what do I need to do to resume the topic, I would like the device to work fully.Have you managed to pull out all entities from this device including total kW.h and Hz

**Reply by make-all** [reply]
Your previous question was deleted because it was about a different integration. I also hid your comment on another bug report on the assumption it was also about the same integration.

 If you are in fact using tuya-local, and having issues, do not post logs or version info from localtuya, as it just causes confusion to discuss issues with other integrations here.

**Comment by sergt1978** [comment]
I contacted the manufacturer of my device, I was informed that this data is not part of a paid subscription. I am asking for your help to expand the integration functions to output information about power (kWh), total active power (kW), power factor, frequency (Hz)
I assume that additional information can be found on the website.
[https://www.zemismart.com/products/sdm01-tw0-12-zm](url)

**Reply by make-all** [reply]
This device is already fully supported by this integration.

**Reply by sergt1978** [reply]
Thanks, then I went to read the instructions.

**Reply by sergt1978** [reply]
I have a wifi boiler, and I haven't updated it, but I still can't log into my device locally. everything works in the cloud. I have a device id and a local key. The IP address and device id match the gifts from the application and router data. The integration sees the device. I enter the local key from the application and the name of the new device is "boiler". The integration still won't let me add it. "Failed to authenticate with device. Verify that the device id and local key are correct.". Replacing the protocol does not do anything

**Reply by sergt1978** [reply]
and this is one of my devices, I'm the same with the others, I have all the data, but the integration does not accept it. yesterday there was an update of two devices, it may not work because of this, I read several entries with a similar reason.

**Reply by fastvd** [reply]
> I contacted the manufacturer of my device, I was informed that this data is not part of a paid subscription. I am asking for your help to expand the integration functions to output information about power (kWh), total active power (kW), power factor, frequency (Hz) I assume that additional information can be found on the website. [https://www.zemismart.com/products/sdm01-tw0-12-zm](url)

<img width="370" height="443" alt="image" src="https://github.com/user-attachments/assets/75c04c7a-860a-4594-b66d-2ac288efbb01" />

Hello. No, you are wrong. The device is still not supported by Tuya Local. If I leave the protocol type as auto when adding a device, it does not correctly assign the device type to me. I also tried all the other protocols... none of them see the device correctly!

**Comment by make-all** [comment]
Discussion #522 is the appropriate topic here.
